torch_np/_detail/implementations.py

```diff
@@ -95,6 +225,129 @@ def diff(a_tensor, n=1, axis=-1, prepend_tensor=None, append_tensor=None):
     return result
 
 
+# #### concatenate and relatives
+
+
+def concatenate(tensors, axis=0, out=None, dtype=None, casting="same_kind"):
+    # np.concatenate ravels if axis=None
+    tensors, axis = _util.axis_none_ravel(*tensors, axis=axis)
+
+    if out is not None or dtype is not None:
+        # figure out the type of the inputs and outputs
+        out_dtype = out.dtype.torch_dtype if dtype is None else dtype
+
+        # cast input arrays if necessary; do not broadcast them agains `out`
+        tensors = _util.cast_dont_broadcast(tensors, out_dtype, casting)
+
+    try:
+        result = torch.cat(tensors, axis)
+    except (IndexError, RuntimeError):
+        raise _util.AxisError
+
+    return result
+
+
+def stack(tensors, axis=0, out=None, *, dtype=None, casting="same_kind"):
+    shapes = {t.shape for t in tensors}
+    if len(shapes) != 1:
+        raise ValueError("all input arrays must have the same shape")
+
+    result_ndim = tensors[0].ndim + 1
+    axis = _util.normalize_axis_index(axis, result_ndim)
+
+    sl = (slice(None),) * axis + (None,)
+    expanded_tensors = [tensor[sl] for tensor in tensors]
+    result = concatenate(
+        expanded_tensors, axis=axis, out=out, dtype=dtype, casting=casting
+    )
+
+    return result
+
+
+def column_stack(tensors_, *, dtype=None, casting="same_kind"):
```

Comment:
In general, I'd suggest implementing all these in terms of `torch.stack`, as it's their closest cousin, and it's less general than `torch.cat` (i.e. more opportunities for optimising it).

Reply:
stack seems to be slower? https://stackoverflow.com/questions/54307225/whats-the-difference-between-torch-stack-and-torch-cat-functions

The way these are written at the moment is a straightforward vendor of what numpy does. So was straighforward for me at least. 
Is this SO answer still relevant?

Reply:
sorry, but where in that question says that stack is slower?

Something that's relevant here is what one of the comment says, that stack can be implemented using `cat` + views, which is the point I made above.

Reply:
The answer by _casualcausality_, from March 8, 2022:

```
import torch
torch.__version__ # 1.10.2
x = [torch.randn(1) for _ in range(10000)]
torch.cat(x).shape, torch.stack(x).shape # torch.Size([10000]), torch.Size([10000, 1])

%timeit torch.cat(x) # 1.5 ms ± 476 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit torch.cat(x).reshape(-1,1) # 1.95 ms ± 534 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit torch.stack(x) # 5.36 ms ± 643 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
I think that things have got better in the last few versions:
```python
In [1]: import torch
   ...: torch.__version__ # 1.10.2
   ...: x = [torch.randn(1) for _ in range(10000)]

In [2]: %timeit torch.cat(x)
918 µs ± 1.87 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [3]: %timeit torch.cat(x).reshape(-1,1)
   ...: oops each)
917 µs ± 19.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [4]: %timeit torch.stack(x)
947 µs ± 2.65 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
so I wouldn't worry about it.

Reply:
Okay, https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/52/commits/6970fd3efdfa0044a06431db03b7922b0c32b577 make all these *stack variants delegate to their eponymous pytorch equivalents. I guess it's the best we can do in the wrapper layer.